### PR TITLE
Refact proposal pallet

### PR DESCRIPTION
**Pull request by riusricardo**

### CODE REVIEW

File: runtime/rococo/src/propose_parachain.rs [hunk 1]
```diff
@@ -200,11 +204,17 @@ decl_module! {
 			);
 			ensure!(validation_code.0.starts_with(runtime_common::WASM_MAGIC), Error::<T>::DefinitelyNotWasm);
 
-			let active_validators = Session::<T>::validators();
+			let mut active_validators = Session::<T>::validators();
+			ValidatorsToRetire::<T>::take().iter().for_each(|v| {
```
Comment (bkchr) [inline runtime/rococo/src/propose_parachain.rs:208]: You take them here and don't put the validators to retire back.

**Review by bkchr — APPROVED** [review]

File: runtime/rococo/src/propose_parachain.rs [hunk 2]
```diff
@@ -338,6 +363,10 @@ impl<T: Config> pallet_session::SessionManager<T::ValidatorId> for Module<T> {
 			validators.extend(proposal.validators);
 		}
 
+		let queued = ValidatorsToAdd::<T>::take();
+		validators.extend(queued);
+		validators.dedup();
```
Comment (ordian) [inline runtime/rococo/src/propose_parachain.rs:368]: dedup only works on sorted sets, I think you need to call `validators.sort();` before
```suggestion
		validators.sort();
		validators.dedup();
```

File: runtime/rococo/src/propose_parachain.rs [hunk 3]
```diff
@@ -289,6 +300,19 @@ decl_module! {
 
 			pallet_balances::Module::<T>::unreserve(&info.proposer, T::ProposeDeposit::get());
 		}
+
+		/// Add a new validator to the set.
+		#[weight = 100_000]
+		fn register_validator(
```
Comment (ordian) [inline runtime/rococo/src/propose_parachain.rs:306]: this should be `register_validators` (plural)
```suggestion
		/// Add new validators to the set.
		#[weight = 100_000]
		fn register_validators(
```

File: runtime/rococo/src/propose_parachain.rs (hunk 1 above)
Comment (ordian) [reply runtime/rococo/src/propose_parachain.rs:208]: Why is that correct? The retirement takes place on new session. Could you elaborate why we take it here?

File: runtime/rococo/src/propose_parachain.rs [hunk 4]
```diff
@@ -100,13 +98,15 @@ struct RegisteredParachainInfo<AccountId, ValidatorId> {
...
-	pub enum Event {
+	pub enum Event<T> where ValidatorId = <T as pallet_session::Config>::ValidatorId {
 		/// A parachain was proposed for registration.
 		ParachainProposed(Vec<u8>, ParaId),
 		/// A parachain was approved and is scheduled for being activated.
 		ParachainApproved(ParaId),
 		/// A parachain was registered and is now running.
 		ParachainRegistered(ParaId),
+		/// A new validator was added to the set.
+		ValidatorRegistered(Vec<ValidatorId>),
```
Comment (ordian) [inline runtime/rococo/src/propose_parachain.rs:109]:
```suggestion
		/// New validators were added to the set.
		ValidatorsRegistered(Vec<ValidatorId>),
```

File: runtime/rococo/src/propose_parachain.rs [hunk 5]
```diff
@@ -289,6 +300,19 @@ decl_module! {
...
+		#[weight = 100_000]
+		fn register_validator(
+			origin,
+			validators: Vec<T::ValidatorId>,
+		) {
+			T::PriviledgedOrigin::ensure_origin(origin)?;
+
+			validators.clone().into_iter().for_each(|v| ValidatorsToAdd::<T>::append(v));
+
+			Self::deposit_event(RawEvent::ValidatorRegistered(validators));
```
Comment (ordian) [inline runtime/rococo/src/propose_parachain.rs:314]:
```suggestion
			Self::deposit_event(RawEvent::ValidatorsRegistered(validators));
```

**Pull request merged**